### PR TITLE
perf: improve sampling/mask/softmax performance (part 1/2)

### PR DESCRIPTION
**Pull request by yzh119**
## 📌 Description

This is the first part of the performance improvement PR for sampling/mask/softmax operator, in this PR, we defer the cross thread reduction till the end of the loop (similar to how FA2 handles denominator) to reduce the number of shuffling and thread sync instructions.

For the second part of the PR, we will implement the Radix TopK algorithm to improve top-k mask logits performance when K is small.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive benchmarking suite for sampling and softmax operations with performance comparison and visualization tools.

* **Chores**
  * Optimized internal kernel execution strategies for improved performance efficiency.

### CODE REVIEW

**Review by bkryu — APPROVED** [review]
LGTM and unit tests are passing and seems like the reduced number of synchronizations are applied throughout correctly.

Nothing to do for this PR, but I was checking if there are any further optimizations to do and came up with #2058 as a followup for completeness.
